src/replication.c

```diff
@@ -3412,6 +3396,233 @@ void dualChannelSetupMainConnForPsync(connection *conn) {
+            return C_ERR;
+        }
+        sdsfree(*err);
+        *err = NULL;
+    }
+    return C_OK;
+}
+
+int syncWithPrimaryHandleReceivePortReplyState(connection *conn, sds *err) {
+    *err = receiveSynchronousResponse(conn);
+    if (*err == NULL) return C_ERR;
+    /* Ignore the error if any, not all the Redis OSS versions support
+     * REPLCONF listening-port. */
+    if (*err[0] == '-') {
+        serverLog(LL_NOTICE,
+                  "(Non critical) Primary does not understand "
+                  "REPLCONF listening-port: %s",
+                  *err);
+    }
+    sdsfree(*err);
+    *err = NULL;
+    return C_OK;
+}
+
+int syncWithPrimaryHandleReceiveIPReplyState(connection *conn, sds *err) {
+    if (server.replica_announce_ip) {
+        *err = receiveSynchronousResponse(conn);
+        if (*err == NULL) return C_ERR;
+        /* Ignore the error if any, not all the Redis OSS versions support
+         * REPLCONF ip-address. */
+        if (*err[0] == '-') {
+            serverLog(LL_NOTICE,
+                      "(Non critical) Primary does not understand "
+                      "REPLCONF ip-address: %s",
+                      *err);
+        }
+        sdsfree(*err);
+        *err = NULL;
+    }
+    return C_OK;
+}
+
+int syncWithPrimaryHandleReceiveCapaReplyState(connection *conn, sds *err) {
+    *err = receiveSynchronousResponse(conn);
+    if (*err == NULL) return C_ERR;
+    /* Ignore the error if any, not all the Redis OSS versions support
+     * REPLCONF capa. */
+    if (*err[0] == '-') {
+        serverLog(LL_NOTICE,
+                  "(Non critical) Primary does not understand "
+                  "REPLCONF capa: %s",
+                  *err);
+    }
+    sdsfree(*err);
+    *err = NULL;
+    return C_OK;
+}
+
+int syncWithPrimaryHandleReceiveVersionReplyState(connection *conn, sds *err) {
+    *err = receiveSynchronousResponse(conn);
+    if (*err == NULL) return C_ERR;
+    /* Ignore the error if any. Valkey >= 8 supports REPLCONF VERSION. */
+    if (*err[0] == '-') {
+        serverLog(LL_NOTICE,
+                  "(Non critical) Primary does not understand "
+                  "REPLCONF VERSION: %s",
+                  *err);
+    }
+    sdsfree(*err);
+    *err = NULL;
+    return C_OK;
+}
+
+int syncWithPrimaryHandleSendPsyncState(connection *conn, sds *err) {
+    if (replicaSendPsyncCommand(conn) == PSYNC_WRITE_ERROR) {
+        *err = sdsnew("Write error sending the PSYNC command.");
+        abortFailover(*err);
+        return C_ERR;
+    }
+    return C_OK;
+}
+
+void syncWithPrimaryHandleError(connection **conn) {
+    connClose(*conn);
+    *conn = NULL;
+    server.repl_transfer_s = NULL;
+    if (server.repl_rdb_transfer_s) {
+        connClose(server.repl_rdb_transfer_s);
+        server.repl_rdb_transfer_s = NULL;
+    }
+    replicationAbortSyncTransfer();
+    server.repl_state = REPL_STATE_CONNECT;
+}
+
+void syncWithPrimaryHandleWriteError(connection *conn, sds *err) {
```

Comment:
the connection pointer needs to be null'd out as well since the object is destructed by `syncWithPrimaryHandleError` in this function.
```suggestion
void syncWithPrimaryHandleWriteError(connection **conn, sds *err) {
```

Reply:
Done